add a simple test to log package

brenol opened the pull request:
we can improve this further if we can add some kind of synchronization to log messages coming to the server, so we can be 100% sure of the order they're arriving.

this is the most simple implementation: it simply asserts how many log messages arrived, based on the methods called. this is ready to be further improved so we can read the messages as if we were a gelf server.

carlosljr commented on api/log/log_test.go:
```diff
@@ -0,0 +1,83 @@
+			break
+		}
+	}
+	if gotMsgs != msgCount {
```
This condition contradicts the one in line 43. It will never happen and tests will always pass.

brenol replied:
Oh yes, thats true! I did not notice that it doesn't make sense *at all*! Sorry for this one, I'll fix it asap.

brenol commented:
Great feedback, thanks!

> It seems that you are trying to test a dependency of huskyCI that is out of the scope of its main role. Why do you think is necessary to simulate a gelf server to tests the log messages for synchronization? It seems that is a concern related to gelf implementation itself and not used for CI environment. Besides, the test doesn't evaluate any logic consistency the way it was implemented. You implemented a function that simulates a gelf server and, in this logic, tests will always pass and there is nothing to evaluate logic consistency inside this function.

Mainly this was done to close #241, unless I mistaken and there is another 'log' package.

I implemented it this way (with a 'simulated gelf server') so we can improve and validate both the development flow (config.DevEnv = true with a ExampleInitLog, ExampleInfo and so on) and the webserver (with a addr running locally so we can see messages going to a server and validating if development env is set).

I will kindly improve on this PR, as long as I'm sure this is a desired "feature", even though it does not improve the project itself in its core.

carlosljr commented:
Hi Breno!

Thanks for your reply!

Now I understand your point!

The issue you are trying to solve is related to tests in log package.

However, for unit tests in this package, you need to do a refactoring in the functions you are trying to test (`Info`, `Warning` e `Error`) so you can make dependency injection in external functions that are called inside those functions.  Like the `SendLog` function from `glbgelf` package and `ErrorGlbgelf`, they have your on logic that needs to be isolated from the functions you want to test.

We have some examples in HuskyCI like in `auth` package. We have `IsValidUser` function to be tested. You can see that other functions called inside that one is implemented in a way so we can inject a fake function that will return a value that we have control. With this, we can test the `IsValidUser`'s behavior. The function dependency injections occurs in `GetPassFromDB` and `GetHashedPass` functions. These functions is defined in an `interface` type called `ClientHandler`, allowing different implementations through composition, e.g. the `FakeClient` struct defined in the test that implements those two functions. So, we injected those functions inside `MongoBasic` struct through `ClientHandler` interface that is a parameter of this struct.

I suggest you to try to improve the tests using the example explained above. For each test, we need to cover a portion of code of your function. In log package it will be easy as long functions will have one condition. So, you will test when `SendLog` returns an `error` or a `nil` value.

For the case where `isDev` is set to `false`, it doesn't matter, because the gelf server behavior is not part of the scope of those tests. But, it was interesting your function that simulates a gelf server. Very good! :)

I'll look forward for those improvements. If you need any help just send me a message.

brenol commented:
Hi @carlosljr,

No problem. I did that (with the gelf server) to modify the package as little as possible, as I wasn't sure if there would be any issues with that kind of refactoring. I'm all in for those changes, however, I try to keep compatibility with the source the most I can, which is why I wen with the fake gelf server at first.

I just modified this PR and went with the safest route I saw, keeping the compatibility with the current  methods: I added a global Logger variable to the log package.

I don't think there is much problem in this case, as huskyCI is mostly a CLI.

I modified all tests to match that new interface

carlosljr commented:
Well done, Breno!

Thanks for your contribution! :)
